contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,187 @@
+
+## Use cases
+
+* Logging agents such as fluentd. Deploy fluentd as DaemonSet and applications only need to communicate with the fluentd in the same node.
+* For a sharded service that keeps per-node local information in each shard.
+* Authenticating proxies such as [aws-es-proxy](https://github.com/kopeio/aws-es-proxy).
+* In container identity wg, being able to give daemonset pods a unique identity per host is on the 2018 plan, and ensuring local pods can communicate to local node services securely is a key goal there. -- from @smarterclayton
+* Regional data costs in multi-AZ setup - for instance, in AWS, with a multi-AZ setup, half of the traffic will switch AZ, incurring regional data Transfer costs, whereas if something was local, it wouldn't hit the network.
+* Users want a priority-based topology aware routing service. For example, want to visit a "same-zone" service,  of course, a "same-host" backend is more preferred.
+
+## Background
+
+It has been a very often requested feature because people need to create a service which connects to local backend Pods, the so called "only-local" service. In this way, they can reduce network latency, improve security, save money and so on. However, because topology is arbitrary, zone, region, rack, generator, whatever, who knows? We should allow arbitrary locality.
+
+`ExternalTrafficPolicy` was added in v1.4, but only for NodePort and external LB traffic. NodeName was added to `EndpointAddress` to allow kube-proxy to filter local endpoints for various future purposes.
+
+Based on our experience of advanced routing setup and recent demo of enabling this in Kubernetes service, this document would like to introduce a more generic way to enable "local-only service".
+
+## Proposal
+
+This proposal builds off of earlier requests to [use local pods only for kube-proxy loadbalancing](https://github.com/kubernetes/kubernetes/issues/7433) and [node-local service proposal](https://github.com/kubernetes/kubernetes/pull/28637). But, this document proposes that not only the particular "node-local" user case should be taken care, but also a more generic way should be figured out.
+
+## Implementation details
+
+### API changes
+
+#### Service API changes
+
+As with service topology, there are currently two types of topology, called `Required` and `Preferred` which denote “hard” vs. “soft” requirements. An example of `required` topology would be “only visit service backends in the same zone” and an example `preferred` topology would be “visit service in the same host is better”. A hard requirement for "same-host" wouldn’t make sense, since you probably have no service backend in some hosts.
+
+```go
+type Service struct {
+  Spec:
+    // If the topology requirements specified by this field are not met, kube-proxy will
+    // not pick endpoints for the service.
+    // ordering doesn't matter, it's an "AND" semantic in Required array.
+    // +optional
+    Required []TopologyTerm
```

Comment:
This seems very complex - what evidence do we have for this need?

I think you're trying to stcik close to pod affinity, but I don't think we need that for this - convince me?

What use cases do we see for something more complex than:

```
    TopologyMode ServiceTopologyMode
    TopologyKey string
```

or maybe:

```
    Topology ServiceTopology
    // ...
  }
  type ServiceTopology struct {
      Mode ServicetopologyMode
      Key string
  }
```

Reply:
Let me try :)

1) I want to set probability for each endpoint, that's why I introduce `Weight` in `Preferred`. With the weight value, I can calculate probabilities accurately. For example, there are 100 endpoints in the same zone, how to make choice? RANDOMLY select one or ALWAYS select a fixed one? Is the probability of service routing not important?

2) As to `ServicetopologyMode`, valid values are ["ignored", "preferred", "required"]? How much preferred can users express? Or, just forget "preferred" and only select values from ["ignored", "required"]? Then "ignored" is the current behavior and "required" is the hard requirement? Users may want a more flexible option.

3) Both `Required` and `Preferred` I proposed are slice, it means that we have multiple choices on topology keys. For example, I want to visit service backend "in same zone" AND "deployed in physical machines". Sorry, I don't know how to express it using a single `ServiceTopology` struct.

Reply:
Also, 

I feel like option 2 is a good start point if we think:
a)	“rr” proxy algorithm is enough.
b)	One filter condition is enough.